src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java

```diff
@@ -258,15 +267,32 @@ private String convertEnumValue(IFieldStructure fieldStructure, String value) {
 
     private Object processComplex(Value value, IFieldStructure fieldStructure) {
         if (fieldStructure.isCollection()) {
+            checkKind(value, fieldStructure.getName(), KindCase.LIST_VALUE);
             return convertComplexList(value.getListValue(), fieldStructure);
         }
+        checkKind(value, fieldStructure.getName(), KindCase.MESSAGE_VALUE);
         return convertByDictionary(value.getMessageValue().getFieldsMap(), fieldStructure);
     }
 
+    private void checkKind(Value value, String fieldName, KindCase expected) {
+        if (value.getKindCase() != expected) {
+            throw new IllegalArgumentException(String.format("Expected '%s' value but got '%s' for field '%s'", expected, value.getKindCase(), fieldName));
+        }
+    }
+
     private List<IMessage> convertComplexList(ListValue listValue, IFieldStructure fieldStructure) {
-        return listValue.getValuesList()
-                .stream()
-                .map(value -> convertByDictionary(value.getMessageValue(), fieldStructure.getReferenceName()))
-                .collect(Collectors.toList());
+        List<Value> valuesList = listValue.getValuesList();
+        int size = valuesList.size();
+        List<IMessage> result = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+            Value value = valuesList.get(i);
+            try {
+                checkKind(value, fieldStructure.getName(), KindCase.MESSAGE_VALUE);
+                result.add(convertByDictionary(value.getMessageValue().getFieldsMap(), fieldStructure));
+            } catch (RuntimeException e) {
+                throw new MessageConvertException("[" + i + "]", e);
```

Comment:
Maybe remove square brackets?

Reply:
They help to clarify the path in the result Exception message.

Reply:
But a number already implies that it's an index (at least for me)

Reply:
I think some users might think that the number is a field name. Like in this case `Msg.field.0`. It is quite easy to confuse the zero with `O` sometimes.
Also, the users who don't know that the field name cannot be a number might think that the number is the actual field name. Of cause, they might think the same about the `[<number>]` but this option still should be more clear than just a number

Reply:
Yes, I'm aggry with @OptimumOpium the `[<number>]` format is more usual to specify the index of a list than strip number.